File: Laba6/Laba6/Models/Task6Class.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;

namespace Laba6.Models
{
    public abstract class Plant
    {
        public string Name { get; set; }
        public string Type { get; set; }
        public bool OnTheVergeOfExtinction { get; set; }

        public abstract void Grow();
        public abstract void Blossom();

        public void Photosynthesize()
        {
            MessageBox.Show($"{this.Name} фотосинтезує");
        }

        public override string ToString()
        {
            return $"Name:{Name} Type:{Type} In Red List:{OnTheVergeOfExtinction}";
        }
    }

    public class Tree : Plant
    {
        public bool LeavesFallInAutumn { get; set; }

        public override void Grow()
        {

        }

        public override void Blossom()
        {

        }

        public override string ToString()
        {
            return $"{base.ToString()} Are leaves fall in autumn:{LeavesFallInAutumn}";
        }
    }

    public class Flowers : Plant
    {
        public string Color { get; set; }

        public override void Grow()
        {

        }

        public override void Blossom()
        {

        }

        public override string ToString()
        {
            return $"{base.ToString()} Color:{Color}";
        }
    }

    public class PlantDatabase
    {
        private List<Plant> _plants = new List<Plant>();

        public void AddPlant(Plant plant)
        {
            _plants.Add(plant);
        }

        public void DisplayInformationOnScreen(Label label)
        {
            var resultText = string.Join(" ", _plants.Select(p => p.ToString() + "\n").ToList());

            label.Content = resultText;
        }

        public void DisplayPlantsOnTheVergeOfExtinction(Label label)
        {
            var resultText = string.Join(" ", _plants.Where(p => p.OnTheVergeOfExtinction).Select(p => p.ToString() + "\n").ToList());

            label.Content = resultText;
        }

        public List<Plant> FindEndangeredSpecies()
        {
            return _plants.Where(p => p.GetType() == typeof(Tree) || p.GetType() == typeof(Flowers)).ToList();
        }

        public void GeneratePlants()
        {
            if (_plants.Count != 0)
            {
                return;
            }

            // Create a list to hold plants
            List<Plant> plantsList = new List<Plant>();

            Random r = new Random();

            // Generate 10 trees with unique parameters
            for (int i = 1; i <= 10; i++)
            {
                Tree tree = new Tree
                {
                    Name = $"Tree{i}",
                    Type = "Deciduous",
                    LeavesFallInAutumn = i % 2 == 0 // Example: Set LeavesFallInAutumn based on even/odd index
                };

                tree.OnTheVergeOfExtinction = r.Next(10) % 2 == 0;

                plantsList.Add(tree);
            }

            // Generate 10 flowers with unique parameters
            for (int i = 1; i <= 10; i++)
            {
                Flowers flower = new Flowers
                {
                    Name = $"Flower{i}",
                    Type = "Perennial",
                    Color = i % 2 == 0 ? "Red" : "Blue" // Example: Set color based on even/odd index
                };

                flower.OnTheVergeOfExtinction = r.Next(10) % 2 == 0;

                plantsList.Add(flower);
            }

            _plants = plantsList;
        }
    }

}

File: Laba6-Python/TaskModels/task6Class.py
from PyQt5.QtWidgets import QApplication, QWidget, QVBoxLayout, QLabel, QMessageBox
from abc import ABC, abstractmethod
import random

class IProduct(ABC):
    @abstractmethod
    def install(self):
        pass

    @abstractmethod
    def update(self, label):
        pass


class IDeveloper(ABC):
    @abstractmethod
    def write_code(self):
        pass

    @abstractmethod
    def test_code(self):
        pass


class Plant(IProduct, ABC):
    def __init__(self):
        self.Name = ""
        self.Type = ""
        self.OnTheVergeOfExtinction = False

    @abstractmethod
    def grow(self):
        pass

    @abstractmethod
    def blossom(self):
        pass

    def photosynthesize(self):
        QMessageBox.information(None, "Photosynthesize", f"{self.Name} фотосинтезує")

    def __str__(self):
        return f"Name:{self.Name} Type:{self.Type} In Red List:{self.OnTheVergeOfExtinction}"


class Tree(Plant):
    def __init__(self):
        super().__init__()
        self.LeavesFallInAutumn = False

    def grow(self):
        pass

    def blossom(self):
        pass

    def __str__(self):
        return f"{super().__str__()} Are leaves fall in autumn:{self.LeavesFallInAutumn}"


class Flowers(Plant):
    def __init__(self):
        super().__init__()
        self.Color = ""

    def grow(self):
        pass

    def blossom(self):
        pass

    def __str__(self):
        return f"{super().__str__()} Color:{self.Color}"


class PlantDatabase:
    def __init__(self):
        self._plants = []

    def add_plant(self, plant):
        self._plants.append(plant)

    def display_information_on_screen(self, label):
        result_text = '\n'.join(str(p) for p in self._plants)
        label.setText(result_text)

    def display_plants_on_the_verge_of_extinction(self, label):
        result_text = '\n'.join(str(p) for p in self._plants if p.OnTheVergeOfExtinction)
        label.setText(result_text)

    def find_endangered_species(self):
        return [p for p in self._plants if isinstance(p, (Tree, Flowers))]

    def generate_plants(self):
        if self._plants:
            return

        plants_list = []

        random.seed()

        for i in range(1, 11):
            tree = Tree()
            tree.Name = f"Tree{i}"
            tree.Type = "Deciduous"
            tree.LeavesFallInAutumn = i % 2 == 0
            tree.OnTheVergeOfExtinction = random.choice([True, False])
            plants_list.append(tree)

        for i in range(1, 11):
            flower = Flowers()
            flower.Name = f"Flower{i}"
            flower.Type = "Perennial"
            flower.Color = "Red" if i % 2 == 0 else "Blue"
            flower.OnTheVergeOfExtinction = random.choice([True, False])
            plants_list.append(flower)

        self._plants = plants_list